src/Juvix/Compiler/Core/Extra.hs

```diff
@@ -53,6 +56,24 @@ shift m = umapN go
       Var i idx | idx >= k -> Var i (idx + m)
       _ -> n
 
+captureFreeVars :: HashMap Index (Maybe BinderInfo) -> Node -> Node
+captureFreeVars fv
+  | n == 0 = id
+  | otherwise = mkLambdas infos . mapFreeVars
+  where
+    (indices, infos) = unzip (HashMap.toList fv)
+    n = length fv
+    s :: HashMap Index Index
+    s = HashMap.fromList (zip indices [0 ..])
+    mapFreeVars :: Node -> Node
+    mapFreeVars = dmapN go
+      where
+        go :: Index -> Node -> Maybe Node
+        go k = \case
+          Var i u
+            | Just v <- s ^. at (u - k) -> Just (Var i (v + k))
+          _ -> Nothing
+
```

Comment:
I don't get the point of all the conversions going on here. First, when invoking `captureFreeVars` in `lambdaLiftNode` we convert BinderList into a HashMap by first converting it to a list of pairs, and then in `captureFreeVars` we immediately convert this HashMap to a list of pairs, and then take the indices and convert them to a HashMap. Couldn't this be streamlined somehow, reducing the number of conversions? If I understand correctly, `indices` is the same as `freevars` in `lambdaLiftNode` modulo order. Infos for them can be fetched directly from BinderList, e.g. by `map (\idx -> BL.lookup idx bi) freevars`

Reply:
The point was to guarantee from the type signature that there are no repeated indices in the argument of captureFreeVars. But probably that's not the most reasonable here. I have updated the code to use a list of tuples instead